### PR TITLE
[meteor] Make check an assertion function

### PR DESCRIPTION
**Pull request by ebroder**
Meteor's check is an assertion function, and Match.test returns a type predicate. TypeScript 3.7 is expressive enough to capture this. This lets us reduce duplication in Meteor code (e.g. needing to keep function parameter types and check calls in sync with each other).

Is this a good idea? I think so - I'd find it useful in my code. Is it worth constraining to the newest version of TypeScript? I'm less sure. My project is already on 3.7 so it's not a big deal to me.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [:heavy_check_mark:] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [:heavy_check_mark:] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: check API is documented [here](https://docs.meteor.com/api/check.html)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/meteor/check.d.ts
Comment (ebroder) [inline types/meteor/check.d.ts]: I couldn't figure out a better way to express this - I'm just looking for a type that I can return but nobody else can (otherwise any type trivially satisfies the interface)

File: types/meteor/check.d.ts
Comment (ebroder) [inline types/meteor/check.d.ts]: Most of these constants aren't actually defined on `Match`, so having them declared here is a bug.

**Comment by perbergland** [comment]
Personally, I think moving to TS 3.7 is fine given that both adornis:typescript and meteor 1.8.3 TS use that version.

(And I like the check improvements as well - I'm looking for a chai (expect) version of them with TS support for unit tests)

**Review by andrei-markeev — CHANGES_REQUESTED** [review]
we will have to use 3.7 sooner or later anyways, so no problem here. guys who haven't migrated to it yet can use older versions of typings.

small thing with the Matcher type, otherwise a very good change, thanks a lot! I actually want to use it right away 👍😄

**Comment by alesn** [comment]
I agree on moving to TypeScript 3.7. Typescript itself - whatever version you like - can be used with Meteor. That's what I've been using before the Typescript compiler plugin for Meteor was made available (and that's what I'm still using because I never received a satisfactory response on what actual advantages using the limited Meteor TypeScript compiler plugin - rather than using 'tsc' directly - should bring). So TS 3.7.3 can be used from the latest Meteor itself using the Meteor compiler plugin or TS 3.7.4 (or any other version) can be used when using TS directly and TS brings useful improvements in every new version, so I am for moving to TS 3.7 (or 3.7.3, which is bundled in the latest Meteor compiler plugin).
One thing though, the version of TypeScript required by DefinitelyTyped definitions is specified in index.d.ts, so it should be updated to match the actually required TS version

**Comment by andrei-markeev** [comment]
> the version of TypeScript required by DefinitelyTyped definitions is specified in index.d.ts, so it should be updated

he did update that, check the diff

File: types/meteor/check.d.ts
Comment (ebroder) [reply types/meteor/check.d.ts]: Aha, "type brand" was the magic phrase I wasn't familiar with. Changed to match more common styles based on some looking around.

**Comment by andrei-markeev** [comment]
Hey!

Please roll back the "Minimum" thing. Just "Typescript Version" is already enforcing the version: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#i-want-to-use-features-from-typescript-31-or-above

Run the tests & lint locally before pushing so that you know you didn't ruin the build. Remember we get email notifications about those pushes. No need to hurry 😉

Also, force-pushes are generally not the best idea after your PR was already reviewed. A single additional commit very clearly shows what have changed since last review. When force-pushing, you basically force reviewer to review the whole thing again from scratch. Also, old links no longer work since they point to no longer existing commits...

**Comment by ebroder** [comment]
Sorry about the pushes, I'll stack new commits going forward. I added the minimum thing because the build was failing without it. It seems like a new change to dtslint that just got pushed: https://github.com/microsoft/dtslint/pull/265. The failure reproduces locally for me if I upgrade dtslint.

**Comment by andrei-markeev** [comment]
"Minimum" - very interesting. Seems like a very recent change. Ok let's see if Travis uses same dtslint version and it actually builds :)

> Sorry about the pushes

Np 👍

I also had this habit of force pushing some time ago. Nobody ever told me anything. And then there was a guy in our team at work who would always force push... Imagine 50 files changed, and he force-pushes, and you just don't have any idea anymore what actual changes he made.

**Comment by ebroder** [comment]
I think I caught a bug in that dtslint change - it was intended to accept the old syntax but doesn't. I'm happy to roll back to the old syntax once that gets fixed, but Travis definitely seems happy with the new syntax too (which I imagine will still be supported).

**Comment by andrei-markeev** [comment]
> A definition owner has approved this PR

something strange is definitely going on 😄

**Review by andrei-markeev — APPROVED** [review]
Note for the maintainer: the bot somehow thought that definition owner has approved before anyone has approved, see above 🤔

**Comment by ebroder** [comment]
Thanks for the :eyes:!

**Comment by armanio123** [comment]
@andrei-markeev thanks for mentioning the bot behavior. We'll take a look, in the meanwhile I'm merging this PR.

**Comment by afrokick** [comment]
@ebroder  Does it work for you with this example?

```
Match.OneOf(String, [String]);
```

The error occurred: `Argument of type 'StringConstructor[]' is not assignable to parameter of type 'StringConstructor`

**Comment by ebroder** [comment]
@afrokick Good catch, there was a bug there. I've opened #44561 with a fix. I added a corresponding test from the docs, but feel free to confirm if that fixes the issue for you.

**Pull request merged**
